fix(valheim): fix open portal button

opello opened the pull request:
**Description**
The "Open" link previously opened the STATUS_HTTP_PORT, which serves a
status.json generated by querying the Valheim server's query port.  Now
it opens the SUPERVISOR_HTTP_PORT which presents a status and control
web interface for the various processes running under the supervisor.

This seems much more useful than the blank default and JSON file
available in the other server.

**⚙️ Type of change**

- [ ] ⚙️ Feature/App addition
- [ ] 🪛 Bugfix
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🔃 Refactor of current code

**🧪 How Has This Been Tested?**
```
.github/scripts/build-catalog.sh stable/valheim
scp catalog/stable/valheim/5.0.4/questions.yaml truenas.local:/mnt/data/ix-applications/catalogs/github_com_truecharts_catalog_main/stable/valheim/5.0.3/questions.yaml
```
Then added a new "valheim-test" app based on that locally modified questions.yaml.  The open button behaved as I'd like instead of opening a 404 page.

**📃 Notes:**
There are additional details here:  
https://github.com/lloesche/valheim-server-docker#status-web-server

**✔️ Checklist:**

- [x] ⚖️ My code follows the style guidelines of this project
- [x] 👀 I have performed a self-review of my own code
- [ ] #️⃣ I have commented my code, particularly in hard-to-understand areas
- [ ] 📄 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [ ] 🧪 I have added tests to this description that prove my fix is effective or that my feature works
- [x] ⬆️ I increased versions for any altered app according to semantic versioning

opello committed:
fix(valheim): include the portalLink directly

Move from the Include template to directly including the contents in
advance of changing the default "open" port.

opello committed:
fix(valheim): fix open portal button

The "Open" link previously opened the STATUS_HTTP_PORT, which serves a
status.json generated by querying the Valheim server's query port.  Now
it opens the SUPERVISOR_HTTP_PORT which presents a status and control
web interface for the various processes running under the supervisor.

This seems much more useful than the blank default and JSON file
available in the other server.

opello committed:
fix(valheim): bump minor version

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/truecharts/charts?pullRequest=4929) <br/>All committers have signed the CLA.

xstar97 commented:
removed the option from the question to disable the supervisor service since its now the main service and status is secondary.

opello commented:
> removed the option from the question to disable the supervisor service since its now the main service and status is secondary.

I wasn't sure if this was the right path when it would affect the "health checks" or however that's all plugged together.

xstar97 commented:
> > removed the option from the question to disable the supervisor service since its now the main service and status is secondary.
> 
> I wasn't sure if this was the right path when it would affect the "health checks" or however that's all plugged together.

lets discuss on discord -> https://discord.gg/JfjxEcFb

opello committed:
fix(valheim): remove unnecessary quotes

opello committed:
fix(valheim): remove extra port

This port is no longer required:
  https://valheimbugs.featureupvote.com/suggestions/219538/please-clarify-is-port-2458-needed-for-dedicated-servers

opello committed:
fix(valheim): order variables consistently

Keep the top level boolean values for groups of settings ahead of the
remaining settings.

opello commented:
The only other thing that it seems might be worth adding in this PR is preventing someone from disabling the supervisord HTTP port.  That's a bit of a built-in foot-gun with it still allowed.

stavros-k commented on charts/stable/valheim/questions.yaml:
```diff
@@ -9,205 +9,201 @@ questions:
           schema:
             type: string
-            default: "admin"
+            default: admin
```
```suggestion
            default: ""
```

Let's have those empty so users explicitly set username/pass

stavros-k commented on charts/stable/valheim/questions.yaml:
```diff
@@ -9,205 +9,201 @@ questions:
             required: true
             private: true
-            default: "REPLACETHIS"
+            default: REPLACETHIS
```
```suggestion
            default: ""
```

stavros-k commented on charts/stable/valheim/questions.yaml:
```diff
@@ -9,205 +9,201 @@ questions:
             required: true
             private: true
-            default: "REPLACETHIS"
+            default: REPLACETHIS
```
```suggestion
            default: ""
```

xstar97 commented:
> The only other thing that it seems might be worth adding in this PR is preventing someone from disabling the supervisord HTTP port. That's a bit of a built-in foot-gun with it still allowed.

That's been removed in the gui from my prs We need to reference it still in the values.yaml

You need to pull first to get changes,  it seems like you're resetting every commit i pushed.

stavros-k commented on charts/stable/valheim/questions.yaml:
```diff
@@ -9,205 +9,201 @@ questions:
+          description: Turn on supervisor's http server.
           schema:
             type: boolean
             default: true
```
Both of those can go I guess.
Unless there is a reason someone would need to disable one of them

stavros-k reviewed (changes requested):
I'm a bit hesitant on this in general.

Breaking changes to just alter a button... I'm not a fan

xstar97 commented:
i removed the variable thats set the backup dir in the GUI since we mount the predefined location in values as `/backups`

stavros-k commented on charts/stable/valheim/questions.yaml:
```diff
@@ -9,205 +9,183 @@ questions:
           schema:
             type: string
-            default: "admin"
+            default: ''
```
```suggestion
            default: ""
```

opello replied:
Should these defaults still be removed under the scope of this PR?

stavros-k commented on charts/stable/valheim/questions.yaml:
```diff
@@ -9,205 +9,183 @@ questions:
             required: true
             private: true
-            default: "REPLACETHIS"
+            default: ''
```
```suggestion
            default: ""
```

opello replied:
Is there any kind of style guide or documentation that says "don't provide default credentials" or something along those lines?

Edit: See https://github.com/truecharts/charts/pull/4929#discussion_r1034468771

stavros-k replied:
Nop, REPLACETHIS was added at somepoint to do some migrations, which would be failing without this.
But it's not needed.

I think it's better to have it empty so users dont blindly hit next and then looking around to find the username/password.

stavros-k commented on charts/stable/valheim/questions.yaml:
```diff
@@ -9,205 +9,183 @@ questions:
             required: true
             private: true
-            default: "REPLACETHIS"
+            default: ''
```
```suggestion
            default: ""
```

opello replied:
Is there any kind of style guide or documentation of conventions for this, and similar things?  (This one was single vs. double quotes for the YAML string.)

stavros-k replied:
Nop, but YAML mostly uses double quotes and single quotes when it's necessary.
Nothing set in stone.

xstar97 replied:
I use single quotes when string is empty so i don't leave it without any quotes and it fails the cli.

opello commented:
> You need to pull first to get changes, it seems like you're resetting every commit i pushed.

I'm well aware, I specifically rewrote and reorganized the work.  I thought that would be obvious from the history I pushed.

opello commented:
> Breaking changes to just alter a button... I'm not a fan

I'm not sure what this means.  What exactly is the breaking change?  I kind of scope-crept the work here to "fix up" the chart.  It seemed worth improving a bit if I was spending some time on it.

opello commented on charts/stable/valheim/values.yaml:
```diff
@@ -9,10 +9,10 @@ secretEnv:
   SUPERVISOR_HTTP_PORT: "{{ .Values.service.main.ports.main.port }}"
+  SUPERVISOR_HTTP: true
+  STATUS_HTTP_PORT: "{{ .Values.service.status.ports.status.port }}"
+  STATUS_HTTP: true
```
I don't know why this reordering change is being made.  It seems unnecessary.  I specifically addressed it in 9bf159179b7744c105b9afbe01f8139cb4544bd3 even.

opello commented on charts/stable/valheim/values.yaml:
```diff
@@ -46,11 +46,11 @@ service:
     type: LoadBalancer
     ports:
-      gameserver:
+      game:
```
I don't disagree with this or `query`, but it seems arbitrarily different.  What's the purpose other than "putting your own stamp" on it?  I tried to find consistency in other Steam servers but there wasn't any that I found.

xstar97 replied:
Keep it simple for services... its a steam game and users can infer is a query for steam.

opello commented on charts/stable/valheim/questions.yaml:
```diff
@@ -16,92 +16,74 @@ questions:
             required: true
         - variable: BACKUPS
-          label: BACKUPS
+          label: Backups
```
These all seem like good UI improvements.  I don't know what they should have been.  But this one seems particularly bad to leave a single word.  `Enable Backups` or something?  I'm not sure which is why I left them the variable names...

opello commented on charts/stable/valheim/questions.yaml:
```diff
@@ -16,92 +16,74 @@ questions:
-          label: SERVER_PUBLIC
-          description: Whether the server should be listed in the server browser (true) or not (false).
+          label: Public Server
+          description: Whether the server should be listed in the server browser.
```
Curious about this change too -- I took it straight from the Dockerfile author's repo.  Seems like unnecessary editorializing to change it here.

stavros-k replied:
In this case, I think it's redundant to  say that it's true or false, given that is a boolean field.

opello replied:
While I agree, it's present in two different description elements.  It seems like a bit of a "slippery slope" when representing the behavior of the system this simply presents access to instead of effecting that system itself.

https://github.com/lloesche/valheim-server-docker#environment-variables

stavros-k commented:
> > Breaking changes to just alter a button... I'm not a fan
> 
> I'm not sure what this means. What exactly is the breaking change? I kind of scope-crept the work here to "fix up" the chart. It seemed worth improving a bit if I was spending some time on it.

Well changing service names etc, might be ok in the kubernetes side, but there is a big chance SCALE UI won't be that happy when someone upgrades the app.

Also, people with enabled ingress on the `main` service, which was the status service, suddenly they will have ingress in the supervisor instead.

opello commented:
> > > Breaking changes to just alter a button... I'm not a fan
> > 
> > I'm not sure what this means. What exactly is the breaking change? I kind of scope-crept the work here to "fix up" the chart. It seemed worth improving a bit if I was spending some time on it.
> 
> Well changing service names etc, might be ok in the kubernetes side, but there is a big chance SCALE UI won't be that happy when someone upgrades the app.

Is there a practical concern or a way to measure whether or not the upgrade path is adversely affected?

> Also, people with enabled ingress on the `main` service, which was the status service, suddenly they will have ingress in the supervisor instead.

This makes sense.  It seemed like both would have to change in parallel.  Is that true?  I asked @Xstar97 if the "main" name was special.  If it wasn't, or didn't have to be, the names could be "status" and "supervisor" instead of "main" and "the other thing."  That way the intent would be independent of the order of the things in the YAML.

stavros-k commented:
> Is there a practical concern or a way to measure whether or not the upgrade path is adversely affected?

Experience only, GUI is not something we built, so we don't really know how it will react.
It surely not when a type of a variable changes, eg int to string, string to bool, etc even if the value is "true" for example.
Even if you add an `enum` in a string, it's not very happy.

In this case there is a chance it won't complain, as it's essentially removing one variable and adding another.
But I can't guarantee

> This makes sense. It seemed like both would have to change in parallel. Is that true? I asked @Xstar97 if the "main" name was special. If it wasn't, or didn't have to be, the names could be "status" and "supervisor" instead of "main" and "the other thing." That way the intent would be independent of the order of the things in the YAML.

`main` is special in our common lib. on all objects.

opello commented:
> > This makes sense. It seemed like both would have to change in parallel. Is that true? I asked @Xstar97 if the "main" name was special. If it wasn't, or didn't have to be, the names could be "status" and "supervisor" instead of "main" and "the other thing." That way the intent would be independent of the order of the things in the YAML.
> 
> `main` is special in our common lib. on all objects.

Okay.  Such a convention must provide immense value to be preferred to strong semantic meaning carried by well named items?  Because it seems to open the door for this kind of situation as things get more complex...  Is there a preferable or more idiomatic way forward?  Or "change isn't allowed" is the takeaway?

You can see my first approach:  
668d97372480d5f19f3cc8fac67c0046c91125b5  
2c1088d9d8d76da3dee4d15b80070b6c3dd70720  
1cc67a1c6b5f6c4a4c88050b9f03da6ff74ff05d

It was a much more minimal change.

stavros-k commented:
> > > This makes sense. It seemed like both would have to change in parallel. Is that true? I asked @Xstar97 if the "main" name was special. If it wasn't, or didn't have to be, the names could be "status" and "supervisor" instead of "main" and "the other thing." That way the intent would be independent of the order of the things in the YAML.
> > 
> > 
> > `main` is special in our common lib. on all objects.
> 
> Okay. Such a convention must provide immense value to be preferred to strong semantic meaning carried by well named items? Because it seems to open the door for this kind of situation as things get more complex... Is there a preferable or more idiomatic way forward? Or "change isn't allowed" is the takeaway?
> 
> You can see my first approach: [668d973](https://github.com/truecharts/charts/commit/668d97372480d5f19f3cc8fac67c0046c91125b5) [2c1088d](https://github.com/truecharts/charts/commit/2c1088d9d8d76da3dee4d15b80070b6c3dd70720) [1cc67a1](https://github.com/truecharts/charts/commit/1cc67a1c6b5f6c4a4c88050b9f03da6ff74ff05d)
> 
> It was a much more minimal change.

While the variable approach is not the best, because it won't work when ClusterIP + ingress is setup.
It's ok for now.

Portal configmap will be (re-)worked in the near future, to correctly generate for all services.

So, yea, I'd prefer minimal changes, instead of having floods of tickets for errors on upgrade.

opello commented:
> While the variable approach is not the best, because it won't work when ClusterIP + ingress is setup. It's ok for now.

> So, yea, I'd prefer minimal changes, instead of having floods of tickets for errors on upgrade.

So, which is the future for this PR?  The minimal change but not getting to use the portalLink template or as-is with the reordering and additional clean up and rework?

Or adding an additional button besides "Open" like "Admin" in some others?  If so, what would be the preferred label?

> Portal configmap will be (re-)worked in the near future, to correctly generate for all services.

In this near future world of a re-worked Portal configmap, what does selecting the visible actions on the applications page look like?  Does each service get an entry in a list?  Are they captioned for the service label?  Seems like it would be space constrained...

stavros-k commented:
> In this near future world of a re-worked Portal configmap, what does selecting the visible actions on the applications page look like? Does each service get an entry in a list? Are they captioned for the service label? Seems like it would be space constrained...

It will be just the configmap generated. What buttons will show will be up to the dev to decide.

For now, single button, pointed to the supervisor port.  Is there a valid reason to have button for the status page?

opello commented:
> For now, single button, pointed to the supervisor port. Is there a valid reason to have button for the status page?

Not that I can see.  It certainly is not useful as-is in master.

And how do you want that achieved?  I'm just looking for guidance on the approach that will be palatable at this point, before updating the PR.

stavros-k commented:
> > For now, single button, pointed to the supervisor port. Is there a valid reason to have button for the status page?
> 
> Not that I can see. It certainly is not useful as-is in master.
> 
> And how do you want that achieved? I'm just looking for guidance on the approach that will be palatable at this point, before updating the PR.

https://github.com/truecharts/charts/commit/2c1088d9d8d76da3dee4d15b80070b6c3dd70720

Like this.. It will still not work when ingress is enabled on supervisor service, but that will be fixed later.

opello commented:
> It will still not work when ingress is enabled on supervisor service, but that will be fixed later.

And what exactly is the fix for that?  Is the problem that the name association of "main" "supervisor" and "status" is preserved when ingress is configured?  Or is there some way to avoid the surprise behavior or inconsistency?  Or given that should the major number change (and then maybe we justify the bigger changes) ...?

stavros-k commented:
> > It will still not work when ingress is enabled on supervisor service, but that will be fixed later.
> 
> And what exactly is the fix for that? 

it's the portal configmap generation. common library side.
I'm refactoring the library already.

> Is the problem that the name association of "main" "supervisor" and "status" is preserved when ingress is configured? Or is       > there some way to avoid the surprise behavior or inconsistency? Or given that should the major number change (and then        > maybe we justify the bigger changes) ...?

I don't really see a good reason to have "breaking" changes. Even if you change the names on the services and it works now. That just hides the inability for the portal generator to create configmaps for other services.

Anyway.

opello commented:
> > Is the problem that the name association of "main" "supervisor" and "status" is preserved when ingress is configured? Or is       > there some way to avoid the surprise behavior or inconsistency? Or given that should the major number change (and then        > maybe we justify the bigger changes) ...?
> 
> I don't really see a good reason to have "breaking" changes. Even if you change the names on the services and it works now. That just hides the inability for the portal generator to create configmaps for other services.
> 
> Anyway.

But the use of a "main" is inherently misleading and obfuscating to the intent of other parts of the chart definition.  It makes reasoning about it harder than necessary due to the sentinel name.  That's why I was talking about the value provided by the shared infrastructure must be worth more than the cognitive load required to understand the system.  This is a judgement call for you guys, but as someone that is just approaching it and not finding it documented, it was difficult to grok.

Furthermore, you didn't really describe the issue with ingress.  Is it because of the name "main" or because of the order of things in another list or something altogether different?  I asked because I don't know, and it's just as time expensive to guess when you could describe the problem.  I did ask.

Breaking things is really easily justified when the thing that is broken was not well defined.  I would argue that the "Open" button not behaving like other "Open" buttons (presenting some sort of control surface of the app) is very broken.  So much so that I'm trying to fix it, and to do so in whatever way makes the most sense to be consistent with the other aspects of this environment, aspects of which I am nowhere near as familiar as you or others.

opello commented:
> Like this.. It will still not work when ingress is enabled on supervisor service, but that will be fixed later.

I think in the current form of this PR ingress should not be adversely affected.  Please let me know if that's wrong.

stavros-k commented:
> > Like this.. It will still not work when ingress is enabled on supervisor service, but that will be fixed later.
> 
> I think in the current form of this PR ingress should not be adversely affected. Please let me know if that's wrong.

Ingress is on port 443, the variable won't point to that port. 
So when ingress enabled, it will point to DNS:supervisor_port instead of DNS:443
But, nothing to do now. I'll handle it on the common lib side

opello commented:
> > > Like this.. It will still not work when ingress is enabled on supervisor service, but that will be fixed later.
> >
> > I think in the current form of this PR ingress should not be adversely affected. Please let me know if that's wrong.
> 
> Ingress is on port 443, the variable won't point to that port.  
> So when ingress enabled, it will point to DNS:supervisor_port instead of DNS:443  
> But, nothing to do now. I'll handle it on the common lib side

Okay, but just to be perfectly clear that isn't a change born out of this PR.  It was like that before and like that it remains.  It was unclear as to if you were attributing the fact to this PR or just a general comment about the state of this chart.

Nothing in the container binds to 443 so that doesn't seem like an applicable fix for this chart anyway.

stavros-k commented:
> > > > Like this.. It will still not work when ingress is enabled on supervisor service, but that will be fixed later.
> > > 
> > > 
> > > I think in the current form of this PR ingress should not be adversely affected. Please let me know if that's wrong.
> > 
> > 
> > Ingress is on port 443, the variable won't point to that port.
> > So when ingress enabled, it will point to DNS:supervisor_port instead of DNS:443
> > But, nothing to do now. I'll handle it on the common lib side
> 
> Okay, but just to be perfectly clear that isn't a change born out of this PR. It was like that before and like that it remains. It was unclear as to if you were attributing the fact to this PR or just a general comment about the state of this chart.
> 
> Nothing in the container binds to 443 so that doesn't seem like an applicable fix for this chart anyway.

Ingress (aka Reverse proxy) listens on 443 (or whatever the port is set in traefik).
So yea, when ingress is enabled on supervisor service, it will try to open on https://ingresshost:443

truecharts-admin commented:
This PR is locked to prevent necro-posting on closed PRs. Please create a issue or contact staff on discord if you want to further discuss this
